### DISCUSSION TITLE
Help needed with SSL certificates

### DISCUSSION THREAD

**Discussion by egavazzi** [cjdoris/PythonCall.jl#493] [Q&A] [answered]
I am calling a Python package from Julia using PythonCall. A function in the package is doing an HTTP request to download some data from a website. When I import and run the function from Julia, I get the following error about some SSL certificate issue.
```julia
ERROR: Python: URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)>
```
However, when I run the same function directly from the Python installed in the local conda environment (the .CondaPkg env), everything is working like a charm. Is there something I am doing wrong? Maybe there are some option I should change in PythonCall related to libraries or something? I am a bit desperate right now after trying some solutions for hours without any success so I am hoping someone more competent will be able to help :sweat_smile:

**Comment by cjdoris** [comment]
Only thing that springs to mind is to see if `os.environ` is different in PythonCall versus directly in Python (in the same Conda environment). If so, as a workaround try setting the appropriate environment vars in Julia before loading PythonCall. Let me know.

**Reply by egavazzi** [reply]
I checked `os.environ` in both PythonCall vs directly Python but did not find any big difference. However I think I found what is the source of the issue. The problem seems to come from calling the package HTTP.jl somewhere else in the code base.

Here is a MWE which works fine:
```julia
using PythonCall

request = pyimport("urllib.request")
request.urlopen("https://celestrak.org/SpaceData/SW-All.csv")
``` 

But this one throws me the error about SSL certificates:
```julia
using PythonCall
using HTTP

request = pyimport("urllib.request")
request.urlopen("https://celestrak.org/SpaceData/SW-All.csv")
``` 
Note however that if you run the second code block (the problematic one) in the same Julia session **after** the first one (the working one), everything works fine. It breaks if you run the second code block in a fresh new Julia session.

So loading `urllib` from Python before loading HTTP seems to fix the problem. This works on my machine for example:
```julia
using PythonCall

request = pyimport("urllib.request")

using HTTP

request.urlopen("https://celestrak.org/SpaceData/SW-All.csv")
```


Now put this into a package which uses  `HTTP.jl` in some function and `urllib` in some dependencies imported from python, add some compilation/precompilation magic and you get a really beautiful mess to debug :smile: 

I don't know however if this kind of 'incompatibility' with HTTP.jl is an expected behavior or a bug (and if it is a bug, if I should report it here or to HTTP.jl?)

**Reply by cjdoris** [reply]
HTTP.jl depends on OpenSSL.jl, do you still see the issue if you just import that? Given the issue is triggered just by importing, you may be able to find the cause in an `__init__` function somewhere.

My guess is that both Julia and Python initialise OpenSSL, but initialise it differently in a way which is affected by load order.

**Reply by egavazzi** [reply]
Good intuition, the issue still appears when just importing OpenSSL.jl. I will try to dig into the `__init__` functions.

**Reply by cjdoris** [reply]
It could also be down to the version of libopenssl being loaded on each side, so check if there is a difference there.

**Reply by egavazzi** [reply]
`OpenSSL.version()` gives `OpenSSL 3.0.13 30 Jan 2024` and `ssl.OPENSSL_VERSION` gives
`Python: 'OpenSSL 3.0.13 30 Jan 2024'` so that does not seem to be the issue.

I started looking at the output of `ssl.get_default_verify_paths()` for the different load orders.

Loading ssl from Python first results in
```
Python: DefaultVerifyPaths(
cafile='/home/etienne/Documents/Julia/PythonCall_tests/MWE_debugging_HTTP/.CondaPkg/env/ssl/cert.pem', 
capath=None, 
openssl_cafile_env='SSL_CERT_FILE', 
openssl_cafile='/home/etienne/Documents/Julia/PythonCall_tests/MWE_debugging_HTTP/.CondaPkg/env/ssl/cert.pem', 
openssl_capath_env='SSL_CERT_DIR', 
openssl_capath='/home/etienne/Documents/Julia/PythonCall_tests/MWE_debugging_HTTP/.CondaPkg/env/ssl/certs')
```

Loading ssl from Python after loading OpenSSL.jl results in
```
Python: DefaultVerifyPaths(
cafile=None, 
capath=None, 
openssl_cafile_env='SSL_CERT_FILE', 
openssl_cafile='/workspace/destdir/ssl/cert.pem', 
openssl_capath_env='SSL_CERT_DIR', 
openssl_capath='/workspace/destdir/ssl/certs')

```

**Reply by cjdoris** [reply]
Here's what I get doing things in different orders.

Order 1: Load `ssl` before `OpenSSL`.
```
julia> using PythonCall

julia> @py import ssl

julia> ssl.get_default_verify_paths()
Python: DefaultVerifyPaths(cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', capath=None, openssl_cafile_env='SSL_CERT_FILE', openssl_cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', openssl_capath_env='SSL_CERT_DIR', openssl_capath='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/certs')

julia> ssl.OPENSSL_VERSION
Python: 'OpenSSL 3.3.0 9 Apr 2024'

julia> using OpenSSL

julia> OpenSSL.version()
"OpenSSL 3.0.13 30 Jan 2024"

julia> ssl.get_default_verify_paths()
Python: DefaultVerifyPaths(cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', capath=None, openssl_cafile_env='SSL_CERT_FILE', openssl_cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', openssl_capath_env='SSL_CERT_DIR', openssl_capath='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/certs')

julia> ssl.OPENSSL_VERSION
Python: 'OpenSSL 3.3.0 9 Apr 2024'
```

Order 2: Load `OpenSSL` before `ssl`
```
julia> using OpenSSL

julia> OpenSSL.version()
"OpenSSL 3.0.13 30 Jan 2024"

julia> using PythonCall

julia> @py import ssl

julia> ssl.get_default_verify_paths()
Python: DefaultVerifyPaths(cafile=None, capath=None, openssl_cafile_env='SSL_CERT_FILE', openssl_cafile='/workspace/destdir/ssl/cert.pem', openssl_capath_env='SSL_CERT_DIR', openssl_capath='/workspace/destdir/ssl/certs')

julia> ssl.OPENSSL_VERSION
Python: 'OpenSSL 3.0.13 30 Jan 2024'

julia> OpenSSL.version()
"OpenSSL 3.0.13 30 Jan 2024"
```

Note that `ssl.OPENSSL_VERSION` is different in the two orders. In Order 2, it is the same as `OpenSSL.version()`, suggesting that Python is using the (older) version of libssl already loaded by `OpenSSL` instead of the one bundled with Python. I don't know if being on the older version is a problem.

And the other main thing to note is that the verify paths are indeed different, as you observed. In particular in Order 1 the `cafile` is set whereas in Order 2 it is `None`.

**Reply by cjdoris** [reply]
Calling the underlying C functions that get the cert paths:

First just with OpenSSL:
```
julia> using OpenSSL: libssl

julia> ccall((:X509_get_default_cert_file, libssl), Cstring, ())
Cstring(0x00007f0a593ca018)

julia> unsafe_string(ans)
"/workspace/destdir/ssl/cert.pem"
```

Now loading Python `ssl` first:
```
julia> using PythonCall

julia> @py import ssl

julia> ssl.get_default_verify_paths()
Python: DefaultVerifyPaths(cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', capath=None, openssl_cafile_env='SSL_CERT_FILE', openssl_cafile='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem', openssl_capath_env='SSL_CERT_DIR', openssl_capath='/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/certs')

julia> using OpenSSL: libssl

julia> ccall((:X509_get_default_cert_file, libssl), Cstring, ())
Cstring(0x00007fb6d9e119a8)

julia> unsafe_string(ans)
"/home/cjdoris/.julia/environments/python/.CondaPkg/env/ssl/cert.pem"
```

This is all consistent with the above findings. I guess really it's just showing that the unwanted default paths really are coming from the version of libssl loaded by OpenSSL.

Python ssl is hard-coded to use `X509_get_default_cert_file` to get the cert file, and that path is hard-coded into libssl. Presumably the version of libssl distributed by Python/Conda is set-up to direct to the cert file distributed with Python/Conda, whereas the version of libssl distributed by OpenSSL.jl (namely OpenSSL_jll) does not do this.

**Reply by egavazzi** [reply]
Ah yes I didn't think about trying different orders when checking libssl. I don't know if forcing a newer version could fix the issue, but I found that setting the environment variable `SSL_CERT_FILE` before importing `ssl` (in the case we already loaded `OpenSSL`) makes it work. I set it to the output of `NetworkOptions.ca_roots()`, which according to the docstring returns the path to the file where the root certificates of the system are stored. 

```julia
using OpenSSL
using NetworkOptions
using PythonCall

ENV["SSL_CERT_FILE"] = NetworkOptions.ca_roots()
request = pyimport("urllib.request")
request.urlopen("https://celestrak.org/SpaceData/SW-All.csv")
```

**Reply by egavazzi** [reply]
I didn't see your newest comment when answering. Thank you for your findings and explanation (I am not sure I would have found these things with `X509_get_default_cert_file` myself, or at least it would have taken a lot more time).
However I don't really understand the cert path returned by libssl used by OpenSSL (`/workspace/destdir/ssl/cert.pem`) and why this works with HTTP.jl but not with urllib.

**Reply by cjdoris** [reply, marked as answer]
I think HTTP.jl just doesn't use OpenSSL's defaults at all and uses NetworkOptions instead to find the certs.

There may be some magic PythonCall can do to get this working automagically, for example it could set `SSL_CERT_FILE` as you do above, or could do
```python
ssl.create_default_context.__kwdefaults__["cafile"] = NetworkOptions.ca_roots()
```

**Reply by egavazzi** [reply]
Nice, I didn't know one could set default keyword arguments this way. Both way work indeed. Thank you very much for your help on this issue, I will mark your comment as answer to the question :+1:
